code/modules/client/preferences/preferences.dm

```diff
@@ -0,0 +1,602 @@
+	ROLE_MINOR		= 1,
+	ROLE_PRISONER	= 1,
+	ROLE_GRUE		= 1,
+	DIVERGENTCLONE  = 1,
+)
+
+var/list/nonantag_roles = list(
+	ROLE_BORER        = 1,
+	ROLE_PLANT        = 1,
+	ROLE_PAI          = 1,
+	ROLE_POSIBRAIN    = 1,
+)
+
+var/list/role_wiki=list(
+	ROLE_ALIEN				= "Xenomorph",
+	BLOBOVERMIND			= "Blob",
+	ROLE_BORER				= "Cortical_Borer",
+	CHANGELING				= "Changeling",
+	CULTIST					= "Cult",
+	ROLE_PLANT				= "Dionaea",
+	MALF					= "Guide_to_Malfunction",
+	NUKE_OP					= "Nuclear_Agent",
+	ROLE_PAI				= "Personal_AI",
+	ROLE_POSIBRAIN			= "Guide_to_Silicon_Laws",
+	REV						= "Revolution",
+	TRAITOR					= "Traitor",
+	CHALLENGER				= "Challengers",
+	VAMPIRE					= "Vampire",
+	VOXRAIDER				= "Vox_Raider",
+	WIZARD					= "Wizard",
+	GRINCH					= "Grinch",
+	NINJA					= "Space_Ninja",
+	TIMEAGENT				= "Time_Agent",
+	PULSEDEMON				= "Pulse_Demon",
+	ROLE_MINOR				= "Minor_Roles",
+	ROLE_PRISONER			= "Minor_Roles",
+	ROLE_GRUE				= "Grue",
+	DIVERGENTCLONE			= "Divergent_Clone",
+)
+
+var/list/special_popup_text2num = list(
+	"Only use chat" = SPECIAL_POPUP_DISABLED,
+	"Only use special" = SPECIAL_POPUP_EXCLUSIVE,
+	"Use both chat and special" = SPECIAL_POPUP_USE_BOTH,
+)
+
+var/list/headset_sound_text2num = list(
+	"Disabled" = HEADSET_SOUND_DISABLED,
+	"Transmit Only" = HEADSET_SOUND_TRANSMIT,
+	"All" = HEADSET_SOUND_ALL,
+)
+
+var/const/MAX_SAVE_SLOTS = 16
+
+#define POLLED_LIMIT	100
+
+/datum/preferences
+	var/list/subsections
+	//doohickeys for savefiles
+	var/database/db = ("players2.sqlite")
+	var/path
+
+	// Which character slot
+	var/slot = 1
+	var/list/slot_names = new
+	var/slot_name = ""
+
+	// 0 = character settings, 1 = game preferences
+	var/current_tab = 0
+
+	// Last time the guy saved their prefs
+	var/lastPolled = 0
+
+	var/savefile_version = 0
+
+	// Alist = associative lists. This is a new 516 thing. Woo!
+	var/alist/preference_settings_client = alist()
+	var/alist/preference_settings_character = alist()
+
+	// Don't like hardcoding this but I can't find a way..
+	var/list/organ_data = list()
+
+	// REALLY don't like hardcoding this. Will be for another time
+	var/list/roles = list()
+
+	//non-preference stuff
+	// Shouldn't these by on client?
+	var/last_ip
+	var/last_id
+	var/muted
+
+	//Mob preview
+	var/icon/preview_icon = null
+	var/icon/preview_icon_front = null
+	var/icon/preview_icon_side = null
+	var/preview_background = null
+	var/list/background_options = list("Black", "White", "Tile")
+
+	var/client/client
+	var/saveloaded = 0
+
+/datum/preferences/New(client/C)
+	client=C
+	if(istype(C))
+		init_datums()
+		init_subsections()
+		var/theckey = C.ckey
+		var/thekey = C.key
+		if(!IsGuestKey(thekey))
+			var/load_pref = try_load_preferences(theckey, C.mob)
+			var/default_slot = get_pref(/datum/preference_setting/numerical/default_slot)
+			if(load_pref)
+				to_chat(C, "Successfully loaded preferences.")
+				while(!SS_READY(SShumans))
+					sleep(1)
+				try_load_save_sqlite(theckey, C, default_slot)
```

Comment:
I've changed this into two different procs depending on whether there is already a character or not